src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -40,6 +103,285 @@ public override bool CanConvert(Type typeToConvert)
         /// <returns>The value that was converted.</returns>
         public abstract T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options);
 
+        internal bool TryRead(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options, ref ReadStack state, out T value)
+        {
+            if (ClassType == ClassType.Value)
+            {
+                // A value converter should never be within a continuation.
+                Debug.Assert(!state.IsContinuation);
+
+                // For perf and converter simplicity, handle null here instead of forwarding to the converter.
+                if (reader.TokenType == JsonTokenType.Null && !HandleNullValue)
+                {
+                    value = default!;
+                    return true;
+                }
+
+#if !DEBUG
+                // For performance, only perform validation on internal converters on debug builds.
+                if (IsInternalConverter)
+                {
+                    value = Read(ref reader, typeToConvert, options);
+                }
+                else
+#endif
+                {
+                    state.Current.OriginalPropertyTokenType = reader.TokenType;
+                    state.Current.OriginalPropertyDepth = reader.CurrentDepth;
+                    state.Current.OriginalPropertyBytesConsumed = reader.BytesConsumed;
```

Comment:
Does these `OriginalProperty*` fields need to be in the state? it seems that they are only used in here, can we use locals instead?

Reply:
Thanks. At one point they were needed, but not anymore.